Feature/home work 3 retrofit

novopashinwm opened the pull request.

novopashinwm committed:
Добавил подробный просмотр информации о фильме:
1) Пришлось добавить полей в класс Movie.kt, чтобы соответствовало проекту.
2) Понимаю, что сделал аляписто в movie_details_fragment.xml с LinearLayout, мне нужна более подрбная информация как более аккуратно выводить данные типа таблица.
3) Удалось скруглить кнопку для просмотра фильма, но не удалось скруглить фотки артистов. Наверняка какое-то свойство есть.
4) Как в MockDetailsRepository.kt вызвать getString, чтобы получить доступ к ресурсам - раньше я получал через Resources.getString - сейчас этот метод depricated.
5) Для просмотра артистов пришлось создать свои item в классах и верстке, а также контейнер для хранения этих артистов.

novopashinwm committed:
Исправил ошибки:
1) Вынес в extentions загрузку картинок.
2) Убрал ArtistCardContainer.kt и работу с 2 RecyclerView.
3) Заменил работу с like на контрол CheckBox
4) Изменил табличку на RelativeLayout.

novopashinwm committed:
1) Вынес в ресурсы высоту RecyclerView с артистами.

novopashinwm committed:
Выполнено ДЗ по retrofit
1) Реализован список now_playing, upcoming, popular.
2) Реализован просмотр популярный телешоу.
3) Реализован просмотр детальной информации по фильму.

AndroidStudentClub commented on app/src/main/java/MovieResponse.kt:
```diff
@@ -0,0 +1,42 @@
+    @SerializedName("total_results")
+    val totalResults: Int
+) {
+    data class Dates(
```
А зачем внутри класса другие data классы размещать? Лучше отдельно не внутри скобок. Иначе получается очень долго и много писать будет приходится - типа MovieResponse.Dates - хотя можно просто Dates

novopashinwm replied:
Разместил классы отдельно.

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/data/ArtistData.kt:
```diff
@@ -0,0 +1,49 @@
+    val id: Int
+) {
+
+    data class Cast(
```
Я бы внутри скобок не описывал data классы

novopashinwm replied:
Вынес.

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/data/MoviesResponse.kt:
```diff
@@ -0,0 +1,20 @@
+    @SerializedName("total_results")
+    val totalResults: Int
+) {
+    data class Dates(
```
Уже же был такой класс, зачем его повторять?

novopashinwm replied:
Не совсем понял о каком классе идет речь? О Dates?!

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/data/TVsResponse.kt:
```diff
@@ -0,0 +1,12 @@
+
+import com.google.gson.annotations.SerializedName
+
+data class TVsResponse(
```
MovieResponse и TVsResponse делаают одну и туже задачу и имеют одну и туже структуру. Так что на мой взгляд их дублировать не имеет смысла. А вот если респонс как то изменится то придется в 2 местах менять

novopashinwm replied:
TVsResponse удалил и пользуюсь только классом MovieResponse

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/extensions/ImageExtentions.kt:
```diff
@@ -0,0 +1,10 @@
+
+fun ImageView.loadImage(imageUrl: String) {
+    Picasso.get()
+        .load("https://image.tmdb.org/t/p/w500${imageUrl}")
```
Вот это (https://image.tmdb.org/t/p/w500) можно вынести в build.gradle как с BASE_URL

novopashinwm replied:
Создал в build.gradle :app переменную BASE_URL_IMAGE и использовал ее.

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/network/ApiClient.kt:
```diff
@@ -0,0 +1,18 @@
+    private var retrofit: Retrofit? = null
+    val client: Retrofit?
+        get() {
+            if (retrofit == null) {
```
Такую проверку retrofit == null можно заменить на retrofit?.let{ }

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/network/BaseApi.kt:
```diff
@@ -0,0 +1,7 @@
+
+open class BaseApi {
+    companion object {
+        const val BASE_URL = "https://api.themoviedb.org/3/"
```
В целом, хорошей практикой является вынесение базового урл в build.gradle 

Это может пригодится, когда у нас разные урл (например один для тестового билда/второй для настоящего который в проде)

Вот статья на эту тему [https://medium.com/@ramonrabello/best-practices-for-working-with-different-build-types-c5cd08b1adc6](https://medium.com/@ramonrabello/best-practices-for-working-with-different-build-types-c5cd08b1adc6)Смотреть пункт [#3](https://github.com/dfyz8-g/MoviesApp/pull/3) Way: Using Gradle Build Config Fields

AndroidStudentClub replied:
```
android {
    // defaultConfig goes here
    buildTypes {
        debug {
            buildConfigField "String", 'BASE_URL', "https://dev-api.yourbackend.com"
        }
        staging {
            buildConfigField "String", 'BASE_URL', "https://stg-api.yourbackend.com"
        }
        release {
            buildConfigField "String", 'BASE_URL', "https://api.yourbackend.com"
        }
    }
}
```

AndroidStudentClub replied:
Типа такого

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/network/ChuckApiClient.kt:
```diff
@@ -0,0 +1,35 @@
+
+object ChuckApiClient {
+
+    private const val BASE_URL = "https://api.themoviedb.org/3/"
```
Повторяется урл
Писал выше про build.gradle

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/network/MovieApiClient.kt:
```diff
@@ -1,3 +1,32 @@
+
+object MovieApiClient {
+
+    private const val BASE_URL = "https://api.themoviedb.org/3/"
```
Если значение базового урл одно - то не нужно его дублировать
Обычно выносят в build.gradle и потом везде используют через BuildConfig

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/network/MovieApiClient.kt:
```diff
@@ -1,3 +1,32 @@
+    private var client: OkHttpClient = OkHttpClient.Builder()
+        .addInterceptor(APICustomInterceptor())
+        .addInterceptor(HttpLoggingInterceptor(CustomHttpLogging()).apply {
+            this.level = HttpLoggingInterceptor.Level.BODY
```
Да, все верно, единственное нужно учесть что в текущей реализации даже если приложение будет собрано для релиза в прод - то там тоже будут логи видны, а это плохо.Можно добавить проверку типа
Build.Config!=Release - чтобы в релизной сборке не было видно логов

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/ui/feed/FeedFragment.kt:
```diff
@@ -88,15 +70,48 @@ class FeedFragment : Fragment(R.layout.feed_fragment) {
+                    call: Call<MoviesResponse>,
+                    response: Response<MoviesResponse>
+                ) {
+                    val movies = response.body()!!.results
```
Обычно оператор !! стараются не использовать и для этого можно использовать ?.let

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/ui/feed/FeedFragment.kt:
```diff
@@ -88,15 +70,48 @@ class FeedFragment : Fragment(R.layout.feed_fragment) {
+                }
+
+                override fun onFailure(call: Call<MoviesResponse>, t: Throwable) {
+                    Log.e("ERROR", t.toString())
```
Можно Tiimber.e

novopashinwm replied:
Использовал и не надо тег писать.

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -1,6 +1,60 @@
+        val movie_id = arguments?.getInt("movie_id")
+        MovieApiClient.apiClient.getMovieDetails(movie_id!!).enqueue(object : Callback<MovieDetails> {
+            override fun onResponse(call: Call<MovieDetails>, response: Response<MovieDetails>) {
+                val movie = response!!.body()!!
```
Обычно !! стараются не использовать, можно заменить на ?.let

novopashinwm replied:
Исправил.

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/ui/movie_details/MovieDetailsFragment.kt:
```diff
@@ -1,6 +1,60 @@
+        MovieApiClient.apiClient.getArtists(movie_id!!).enqueue(object : Callback<ArtistData>
+            {
+                override fun onResponse(call: Call<ArtistData>, response: Response<ArtistData>) {
+                    val artists = response!!.body()!!
```
Обычно !! стараются не использовать, можно заменить на ?.let

novopashinwm replied:
Исправил.

AndroidStudentClub commented on app/src/main/java/ru/androidschool/intensiv/ui/tvshows/TvShowsFragment.kt:
```diff
@@ -28,8 +33,17 @@ class TvShowsFragment : Fragment() {
         super.onViewCreated(view, savedInstanceState)
+        MovieApiClient.apiClient.getTVPopular().enqueue(object : Callback<TVsResponse> {
+            override fun onResponse(call: Call<TVsResponse>, response: Response<TVsResponse>) {
+                val tvs = response.body()!!.results
```
Обычно !! стараются не использовать, можно заменить на ?.let

novopashinwm replied:
Исправил.

AndroidStudentClub commented on app/src/main/res/layout/item_card_actors.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Кстати LinearLayout отсюда можно убрать и оставить только RecyclerView

AndroidStudentClub reviewed (approved):
Проверил

novopashinwm committed:
Исправил замечания

1) Основное - убрать лишние классы.
2) Убрать !! и использовать ?.let.
3) Вынес в настройки BASE_URL_IMAGE.
